File: project4/greedyTSP.py
import sys
import math


def greedyTsp(cities, numCities, q):
	minCost = sys.maxint
	cheapesRoute = []
	for startingCity in range(numCities):
		print str(startingCity)
		results = findRoute(startingCity, numCities, cities, minCost)	
		if(results != None):
			totalCost = results[1]
			if(totalCost < minCost):
		    		minCost = totalCost
				if(not q.empty()):
					q.get()	
					q.get()			
				q.put(results[0])
				q.put(minCost)
	return
##END 'def greedyTsp()'



def findRoute(startCity, numCities, cities, minCost):
	
	citiesToCheck = nList(startCity, numCities)
	currentCity = startCity
	route = [(startCity, 0)]
	totalLen = 0
	count = 0
	
	while count < numCities - 1:
		shortest = sys.maxint
		closestCityId = -1
		for id in citiesToCheck:
			length = getLength(cities, id, currentCity)
			if(length < shortest):
				shortest = length		
				closestCityId = id
		
		totalLen += shortest

		#prematurely kill if cost is too high
		if(totalLen > minCost):
			return None
		
		currentCity = closestCityId
		route.append((currentCity, shortest))
		citiesToCheck.remove(currentCity)
		count += 1

	##Route back to start node	
	endLen = getLength(cities, startCity, currentCity)
	totalLen += endLen
	route.append((startCity, endLen))
		
	return route, totalLen
	


def nList(start, n):
	list = []
	for i in range(0, n):
		if(i != start):
			list.append(i)
	return list

	
def getLength(cities, city1, city2):
	city1x = cities[city1][1]
	city1y = cities[city1][2]
	city2x = cities[city2][1]
	city2y = cities[city2][2]
	return int(round(math.sqrt(((city1x - city2x)**2) + ((city1y - city2y)**2))))

	

File: project4/greedyTSP.c
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>


int main(int argc, int argv[]) {
	
	
	int cost(int cities[][2], int city1, int city2) {
		
		int x1 = cities[city1][1];
		int y1 = cities[city1][2];
		int x2 = cities[city2][1];
		int y2 = cities[city2][2];
		
		float a = (x2 - x1)^2;
		float b = (y2 - y1)^2;
		
		return (a + b)^0.5;
	}
	
	int* findRoute(int cities[][2], int numCities, int startCity) {
		
		
		int* currentRoute = (int*) malloc(sizeof(int) * numCities);
		int* cheapestRoute = (int*) malloc(sizeof(int) * numCities);
		int currentCity = startCity;
		int cheapestCity;
		int currentCost = INT_MAX;
		int totalCost = 0;
		int city, i, minCost;
		
		
		int* citiesToCheck = (int*) malloc(sizeof(int) * numCities);
		for (i = 0; i < numCities; i++) {
			citiesToCheck[i] = i;
		}
		
		

		/* Iterate over all the cities */
		for (currentCity = 1; currentCity < numCities; currentCity++) {
			cheapestCity = INT_MAX;
			minCost = INT_MAX;
			
			
			/* Find the city closest to the current city */
			for (city = 0; city < numCities; city++) {
				
				if (citiesToCheck[city] == -1)
					continue;
				
				
				currentCost = cost(cities, currentCity, city);
				
				if (currentCost < minCost) {
					minCost = currentCost;
					totalCost += currentCost;
					cheapestCity = city;
				}
				
				citiesToCheck[city] = -1;
			}
			
			cheapestRoute[0] = totalCost;
			cheapestRoute[currentCity] = cheapestCity;	
		}
		
		return cheapestRoute
	}

		

	/* for each city, cities[][] holds an x and y coordinate */
	int* greedyTsp(int cities[][2], int numCities) {
		
		
		/* The first number in currentRoute is the cost of the route.
			The subsequent numbers represent the sequence of cities visited. */
		int* currentRoute = (int*) malloc(sizeof(int) * numCities);
		int* cheapestRoute = (int*) malloc(sizeof(int) * numCities);
		int minCost = INT_MAX;
		int currentCost;
		int city;
				
		for (city = 0; city < numCities; city++) {
			currentRoute = findRoute(cities, numCities, city);
			currentCost = currentRoute[0];
				
			if (currentCost < minCost) {
				minCost = currentCost;
				cheapestRoute = currentRoute;
			}
		}
			
		return cheapestRoute;
	}
}

